### PR TITLE
Optimize parameter sweeps to cache unparameterized prefixes

### LINKED ISSUE #4208: Optimize parameter sweeps to cache unparameterized prefixes
**Is your feature request related to a use case or problem? Please describe.**
Sweeps rerun the whole circuit from scratch for each parameter set. They could be made more efficient by caching the result of the unparameterized prefix and restarting each parameterized simulation from there.

**Describe the solution you'd like**
We do something similar in `run`, where we cache the result of the unitary prefix, and rerun repetitions starting from there. So it should be straightforward to add this same style of optimization for sweeps.

**Describe alternatives/workarounds you've considered**

* Could do nothing: I assume in practice, parameterized circuits usually have *some* parameters at or near the beginning. Also once we merge the feature that simulates unentangled qubits independently, the first part of the circuit should be a tiny fraction of the simulation time.
* Going the other direction, this could be extended to where if the first N parameters are the same in some parameter sets, it can start each run from the Nth (factoring the circuit would be a fun challenge).
* Tangentially, I wonder if this logic really belongs in the simulator class itself. Seems these could be module-level functions since they just call public simulator methods. This would allow us to create different sweep functions for different optimization styles without having to create a specific simulator class for each.

**What is the urgency from your perspective for this issue? Is it blocking important work?**
P3 - I'm not really blocked by it, it is an idea I'd like to discuss / suggestion based on principle

### PR DESCRIPTION
**Pull request by daxfohl**
Fixes #4208 by simulating the non-parameterized prefix once, and sweeping on the parameterized suffix only.

### CODE REVIEW

File: cirq-core/cirq/sim/simulator_base.py [hunk 1]
```diff
@@ -281,6 +281,55 @@ def _run(
...
+            split_into_matching_protocol_then_general(program, sweep_prefixable)
+            if self._can_be_in_run_prefix(self.noise)
+            else (program[0:0], program)
+        )
+        step_result = None
+        for step_result in self._core_iterator(
+            circuit=prefix,
+            sim_state=act_on_args,
+        ):
+            pass
```
Comment (mpharrigan) [inline cirq-core/cirq/sim/simulator_base.py:329]: what's going on here? are you just trying to get the last `step_result`? is there a more idiomatic way of doing this?

File: cirq-core/cirq/sim/simulator_base.py [hunk 2]
```diff
@@ -281,6 +281,55 @@ def _run(
...
+        act_on_args = step_result._sim_state
+        yield from super().simulate_sweep_iter(suffix, params, qubit_order, act_on_args)
```
Comment (mpharrigan) [inline cirq-core/cirq/sim/simulator_base.py:331]: why is this called `act_on_args`? isn't it a state?

**Review by mpharrigan — COMMENTED** [review]
do you have any examples with example speedups?

File: cirq-core/cirq/sim/simulator_base.py (hunk 2 above)
Comment (daxfohl) [reply cirq-core/cirq/sim/simulator_base.py:331]: Yeah, the class is called ActOnArgs for vestigial reasons but I'll update the variable name here.

File: cirq-core/cirq/sim/simulator_base.py (hunk 1 above)
Comment (daxfohl) [reply cirq-core/cirq/sim/simulator_base.py:329]: Yes, it seems to be the idiomatic way https://stackoverflow.com/questions/2138873/cleanest-way-to-get-last-item-from-python-iterator

**Comment by daxfohl** [comment]
I'm not aware of a specific example, however the issue #4208 was marked as accepted so this PR is based off that. It's a good point though: while it seems obvious this is an improvement because it avoids repeating steps, I'll do a couple perf samples to make sure the wall clock time is actually improved.

**Comment by daxfohl** [comment]
@mpharrigan 

Tested this with the following function:
```python
def test_me():
    circuit = cirq.Circuit()
    for i in range(11):
        for _ in range(9):
            circuit.append(cirq.X(cirq.LineQubit(i)))
        circuit.append(cirq.XPowGate(exponent=sympy.Symbol('a'))(cirq.LineQubit(i)))
        circuit.append(cirq.measure(cirq.LineQubit(i)))

    simulator = cirq.DensityMatrixSimulator(split_untangled_states=False)
    params = [
        cirq.ParamResolver({'a': 0}),
        cirq.ParamResolver({'a': 1}),
        cirq.ParamResolver({'a': 2}),
        cirq.ParamResolver({'a': 3}),
        cirq.ParamResolver({'a': 4}),
    ]

    simulator.simulate_sweep(program=circuit, params=params)
```

Prior to this change, 15.64s. After the change, 6.78s

**Review by mpharrigan — APPROVED** [review]

**Comment by daxfohl** [comment]
@mpharrigan I don't have merge permission. Can you set this to merge?

**Comment by dabacon** [comment]
let's see if I now have requisite super powers

**Pull request merged**